File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.py
def prm(n, num, visited, numbers, number_permutation):
    if len(num) == n:
        number_permutation.add(int(num))
        return
    
    for i in range(0, len(visited)):
        if not visited[i]:
            visited[i] = True
            prm(n, num + numbers[i], visited, numbers, number_permutation)
            visited[i] = False

def solution(numbers):
    answer = 0
    n = len(numbers)
    number_permutation = set()
    visited = [False] * n
    
    # 1개 ~ n개 까지의 종이조각으로 만들 수 있는 순열 생성
    for i in range(1, n + 1):
        prm(i, "", visited, numbers, number_permutation)

    # 0과 1은 소수가 아니다 제외
    number_permutation -= {0, 1}
    
    # 소수인지 확인
    for number in list(number_permutation):
        for j in range(2, int(number ** 0.5) + 1):
            if number % j == 0:
                break
        else:
            answer += 1        
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.HashSet;
import java.util.Set;
import java.util.Arrays;

class Solution {    
    public void permutation(Set<Integer> numSet, String[] numberArray, boolean[] visited, int target, int dept, StringBuilder sb) {
        if (target == dept) {
            numSet.add(Integer.parseInt(sb.toString()));
            return;
        }
        
        for (int idx = 0; idx < numberArray.length; idx++) {
            if (!visited[idx]) {
                sb.append(numberArray[idx]);
                visited[idx] = true;
                permutation(numSet, numberArray, visited, target, dept + 1, sb);
                sb.deleteCharAt(dept);
                visited[idx] = false;
            }
        }
    }
    
    public int solution(String numbers) {
        int answer = 0;
        
        Set<Integer> numSet = new HashSet<Integer>();   // 중복 숫자 제외를 위해 Set 자료형 사용
        StringBuilder sb;
        boolean[] visited;
        // numbers에서 i 개를 뽑았을 때 만들수 있는 숫자 순열 만들기
        for (int i = 1; i <= numbers.length(); i++) {
            sb = new StringBuilder();
            visited = new boolean[numbers.length()];
            permutation(numSet, numbers.split(""), visited, i, 0, sb);
        }
        
        // 만든 숫자 순열을 순회하며 소수가 몇개 인지 확인
        System.out.println(numSet);
        for (int num : numSet) {
            if (num < 2) continue;
            answer++;
            for (int n = 2; n * n <= num; n++) {
                if (num % n == 0) {
                    answer--;
                    break;
                }
            }
        }
        
        return answer;
    }
}